functions.ipynb

```diff
@@ -0,0 +1,1490 @@
+    "- `223450` does **not** meet these criteria (`50` is a decreasing pair of digits)\n",
+    "- `123789` does **not** meet these criteria (no double digit)\n",
+    "\n",
+    "Write a function that determines if a password is valid according to the criteria above.\n",
+    "\n",
+    "**How many valid password are there in your range?**\n",
+    "\n",
+    "*Hint:* the solution function should take **two** arguments: the `start` and `end` of your password range"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "517f4d43-0ffa-407d-9035-7c8e47b64103",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%ipytest\n",
+    "\n",
+    "def solution_password_validator1(start: int, end: int) -> int:\n",
+    "    \"\"\"\n",
+    "    Write your solution here\n",
+    "    \"\"\"\n",
+    "    pass"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "edac5c6a-23fa-4f09-a462-5682201a449f",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "---\n",
+    "\n",
+    "### Part 2 🌶️🌶️"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "af330663-15c2-48f7-b407-e09cecc62d4d",
+   "metadata": {},
+   "source": [
+    "You have a new rule: the two adjacent matching digits **must not be part of a larger group of matching digits**.\n",
+    "\n",
+    "Write a new function `count_valid_p2()` to account for the new rule.\n",
+    "\n",
+    "**How many valid password does your range contain now?**\n",
+    "\n",
+    "---\n",
+    "\n",
+    "A few examples:\n",
+    "\n",
+    "- `112233` meets these criteria because the digits never decrease and all repeated digits are exactly two digits long\n",
+    "- `123444` **doesn't** meet the criteria (the repeated `44` is part of a larger group of `444`)\n",
+    "- `111122` meets the criteria (even though `1` is repeated more than twice, it still contains a double `22`)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "17106c0f-e9df-4888-bb21-e2806cc72df4",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%ipytest\n",
+    "\n",
+    "def solution_password_validator2(start: int, end: int):\n",
+    "    \"\"\"\n",
+    "    Write your solution here\n",
+    "    \"\"\"\n",
+    "    pass"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "503a18e5-8ee4-4a02-89f9-09cbca3a8e50",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "## Buckets reorganization"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "957eb5a0-5331-48e3-86b6-7b99299d710f",
+   "metadata": {},
+   "source": [
+    "### Part 1 🌶️"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "59093ed8-b422-4f97-8ffc-801f6149458a",
+   "metadata": {},
+   "source": [
+    "You have a list of buckets, each containing some items labeled from `a` to `z`, or from `A` to `Z`. Each bucket has two large compartments, and each compartment always has the same number of items.\n",
```

Comment:
Please clarify it a bit. Here is a suggestion:

"You have a list of buckets containing an even number of items that can be any letter from a to Z. Each bucket is split in the middle into two equally sized compartments."

Reply:
This was not modified.

Reply:
```suggestion
    "You have a list of buckets, each containing some items labeled from `a` to `z`, or from `A` to `Z`. Each bucket is split into **two** equally sized compartments.\n",
```